In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [2]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
# YOUR CODE HERE!
file = Path("Resources/lending_data.csv")
lending_df =pd.read_csv(file)
# Review the DataFrame
# YOUR CODE HERE!
lending_df

loan_size  interest_rate  borrower_income  debt_to_income  \
0        10700.0          7.672            52800        0.431818   
1         8400.0          6.692            43600        0.311927   
2         9000.0          6.963            46100        0.349241   
3        10700.0          7.664            52700        0.430740   
4        10800.0          7.698            53000        0.433962   
...          ...            ...              ...             ...   
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
0                    5                 1       22800            0  
1                    3                 0       13600            0  
2                    3                 0       16100            0  
3                    5                 1       22700            0  
4                    5                 1       23000            0  
...                ...               ...         ...          ...  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1  

[77536 rows x 8 columns]

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [3]:
# Separate the data into labels and features


# Separate the y variable, the labels
# YOUR CODE HERE!]
y = lending_df['loan_status']
# Separate the X variable, the features
# YOUR CODE HERE!
x = lending_df.drop(columns='loan_status')

In [4]:
# Review the y variable Series
# YOUR CODE HERE!
y.sample(10)

53650    0
39921    0
42703    0
64327    0
59758    0
73264    0
16722    0
12216    0
17947    0
44531    0
Name: loan_status, dtype: int64

In [5]:
# Review the X variable DataFrame
# YOUR CODE HERE!

x.sample(5)

loan_size  interest_rate  borrower_income  debt_to_income  \
76075    19000.0         11.189            85900        0.650757   
47711     9300.0          7.098            47400        0.367089   
22047     7700.0          6.402            40800        0.264706   
49137     9700.0          7.257            48900        0.386503   
10221     9400.0          7.103            47400        0.367089   

       num_of_accounts  derogatory_marks  total_debt  
76075               12                 2       55900  
47711                3                 0       17400  
22047                2                 0       10800  
49137                4                 0       18900  
10221                3                 0       17400

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [6]:
# Check the balance of our target values
# YOUR CODE HERE!
y.value_counts

<bound method IndexOpsMixin.value_counts of 0        0
1        0
2        0
3        0
4        0
        ..
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, Length: 77536, dtype: int64>

### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [7]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# YOUR CODE HERE!
X_train, X_test, y_train, y_test = train_test_split(x, y, random_state=1, stratify=y)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [8]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
classifier = LogisticRegression(random_state = 1)
# Fit the model using training data
# YOUR CODE HERE!

classifier.fit(X_train, y_train)

LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [9]:
# Make a prediction using the testing data
# YOUR CODE HERE!

predictions = classifier.predict(X_test)
pd.DataFrame({"Prediction": predictions, "Actual": y_test}).tail(10)

Prediction  Actual
76038           1       1
19119           0       0
45193           0       0
25211           0       0
61697           0       0
38069           0       0
36892           0       0
5035            0       0
40821           0       0
35030           0       0

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [10]:
# Print the balanced_accuracy score of the model
# YOUR CODE HERE!

print(f"Training Data Score: {classifier.score(X_train, y_train)}")
print(f"Testing Data Score: {classifier.score(X_test, y_test)}")
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test, predictions)}")

Training Data Score: 0.9914878250103177
Testing Data Score: 0.9924164259182832
Balanced Accuracy Score: 0.9442676901753825


In [11]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!

c_matrix = confusion_matrix(y_test, predictions)
confusion_matrix_df = pd.DataFrame(c_matrix,
                                   index=["Healthy Loans","Non-Healthy Loans"],
                                   columns=["Predicted Healthy Loans", "Predicted Non-Healthy Loans"])
confusion_matrix_df


Predicted Healthy Loans  Predicted Non-Healthy Loans
Healthy Loans                        18679                           80
Non-Healthy Loans                       67                          558

In [12]:
# Print the classification report for the model
# YOUR CODE HERE!

print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      0.89      0.88       625

    accuracy                           0.99     19384
   macro avg       0.94      0.94      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:**The logistic regression model demonstrates strong predictive performance for both the 0 (healthy loan) and 1 (high-risk loan) labels.

For healthy loans (0), the model achieved a perfect prediction rate of 100%. This means that all instances classified as healthy loans were accurately identified by the model.

Regarding high-risk loans (1), the model achieved a prediction rate of 87%. This implies that the model correctly identified 87% of the high-risk loans in the dataset.

The Balanced Accuracy Score, which takes into account the accuracy of both classes, is 0.9442676901753825. This score indicates that the model performs well in capturing the overall patterns and making accurate predictions for both healthy and high-risk loans, considering the class imbalance in the data.

In summary, the logistic regression model exhibits excellent predictive capabilities, achieving a perfect prediction rate for healthy loans and a strong accuracy for high-risk loans. The Balanced Accuracy Score further confirms the model's overall effectiveness in accurately classifying loans into the appropriate categories

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [20]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!
ROS = RandomOverSampler(random_state=1)
ROS


# Fit the original training data to the random_oversampler model
# YOUR CODE HERE!

ROS.fit(X_train, y_train)
X_over, y_over = ROS.fit_resample(X_train, y_train)

In [21]:
# Count the distinct values of the resampled labels data
# YOUR CODE HERE!

y_over.value_counts()

1    56277
0    56277
Name: loan_status, dtype: int64

### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [23]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
# YOUR CODE HERE!

LR = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
# YOUR CODE HERE!

LR.fit(X_over, y_over)

# Make a prediction using the testing data
# YOUR CODE HERE!

LR_prediction = LR.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [24]:
# Print the balanced_accuracy score of the model 
# YOUR CODE HERE!
print(f"Balanced Accuracy Score: {balanced_accuracy_score(y_test, LR_prediction)}")


Balanced Accuracy Score: 0.9959744975744975


In [26]:
# Generate a confusion matrix for the model
# YOUR CODE HERE!
c_matrix = confusion_matrix(y_test, LR_prediction)
confusion_matrix_df = pd.DataFrame(c_matrix,
                                   index=["Healthy Loans","Non-Healthy Loans"],
                                   columns=["Predicted Healthy Loans", "Predicted Non-Healthy Loans"])
confusion_matrix_df


Predicted Healthy Loans  Predicted Non-Healthy Loans
Healthy Loans                        18668                           91
Non-Healthy Loans                        2                          623

In [27]:
# Print the classification report for the model
# YOUR CODE HERE!

print(classification_report(y_test, LR_prediction))


              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18759
           1       0.87      1.00      0.93       625

    accuracy                           1.00     19384
   macro avg       0.94      1.00      0.96     19384
weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** When using logistic regression with oversampled data, the model exhibits excellent performance in predicting both the 0 (healthy loan) and 1 (high-risk loan) labels.

The classification report shows that the model achieved a 100% accuracy in correctly predicting healthy loans (0). This means that all instances classified as healthy loans were indeed labeled correctly.

For high-risk loans (1), the model achieved an 87% accuracy. This implies that the model correctly identified 87% of the high-risk loans in the dataset.

Additionally, the Balanced Accuracy Score is an impressive 0.9959744975744975. This score takes into account the accuracy of both classes, considering the class imbalance in the dataset. With such a high Balanced Accuracy Score, the model demonstrates remarkable success in capturing the patterns and making accurate predictions for both healthy and high-risk loans.

Overall, the logistic regression model, trained on oversampled data, showcases outstanding predictive performance, achieving a near-perfect accuracy for healthy loans and a high accuracy for high-risk loans.






